File: proxy_switch/features/system_proxy.py
"""System-wide proxy environment variables.

Writes proxy settings to /etc/environment and /etc/profile.d/proxy-switch.sh.
Only modifies proxy-related lines — never touches other config in these files.
"""

from __future__ import annotations

from typing import Dict, List

from ..core.models import Result, StatusInfo

NAME = "system_proxy"
DESCRIPTION = "System environment proxy variables (/etc/environment)"
CONFIG_FILES = ["/etc/environment", "/etc/profile.d/proxy-switch.sh"]

ENV_FILE = "/etc/environment"
PROFILE_FILE = "/etc/profile.d/proxy-switch.sh"

_PROXY_KEYS_LOWER = ("http_proxy", "https_proxy", "ftp_proxy", "no_proxy")
_PROXY_KEYS_UPPER = ("HTTP_PROXY", "HTTPS_PROXY", "FTP_PROXY", "NO_PROXY")


def detect(executor=None) -> bool:
    """System proxy can always be applied — every Linux has /etc."""
    return True


def _read_file(executor, path: str) -> str:
    """Read a remote or local file, returning empty string on failure."""
    if executor:
        r = executor.run(f"cat {path} 2>/dev/null || echo ''")
        return r.stdout
    try:
        with open(path) as f:
            return f.read()
    except (FileNotFoundError, PermissionError):
        return ""


def _write_file(executor, path: str, content: str, sudo: bool = False) -> bool:
    """Write content to a remote or local file. Returns True on success."""
    if executor:
        r = executor.write(path, content, sudo=sudo)
        return r.returncode == 0
    import tempfile, subprocess, os
    try:
        with tempfile.NamedTemporaryFile(mode="w", delete=False) as tf:
            tf.write(content)
            tf.flush()
        if sudo:
            r = subprocess.run(["sudo", "cp", tf.name, path],
                               capture_output=True, text=True)
        else:
            import shutil
            shutil.copy(tf.name, path)
            r = subprocess.CompletedProcess([], 0)
        os.unlink(tf.name)
        return r.returncode == 0
    except Exception:
        return False


def _update_env_lines(content: str, proxy_config: Dict[str, str]) -> str:
    """Update only proxy-related lines in /etc/environment content.

    Preserves all non-proxy lines. Adds/replaces proxy lines.
    """
    proxy_prefixes = tuple(
        [k + "=" for k in _PROXY_KEYS_LOWER] +
        [k + "=" for k in _PROXY_KEYS_UPPER] +
        ["# Managed by proxy-switch"]
    )

    # Keep lines that are NOT proxy-related
    keep = [l for l in content.split("\n")
            if not l.strip().startswith(proxy_prefixes)]

    # Build new proxy block
    new_lines = ["# Managed by proxy-switch"]
    for key in _PROXY_KEYS_LOWER:
        val = proxy_config.get(key, "")
        if val:
            new_lines.append(f'{key}="{val}"')
    for key in _PROXY_KEYS_UPPER:
        lower_val = proxy_config.get(key.lower(), "")
        if lower_val:
            new_lines.append(f'{key}="{lower_val}"')

    # Remove trailing blank lines from keep
    while keep and keep[-1] == "":
        keep.pop()

    if keep:
        return "\n".join(keep) + "\n\n" + "\n".join(new_lines) + "\n"
    else:
        return "\n".join(new_lines) + "\n"


def _make_profile_content(proxy_config: Dict[str, str]) -> str:
    """Generate shell profile format — export KEY="value"."""
    lines = ["# Managed by proxy-switch", ""]
    for key in _PROXY_KEYS_LOWER:
        val = proxy_config.get(key, "")
        if val:
            lines.append(f'export {key}="{val}"')
    for key in _PROXY_KEYS_UPPER:
        lower_val = proxy_config.get(key.lower(), "")
        if lower_val:
            lines.append(f'export {key}="{lower_val}"')
    lines.append("")
    return "\n".join(lines)


def enable(proxy_config: Dict[str, str], executor=None) -> Result:
    """Set proxy env vars. Only modifies proxy lines — safe for existing config."""
    # Early sudo check — fail fast instead of hanging on password prompt
    if executor and not executor.has_sudo():
        return Result(
            success=False,
            message="Sudo access required. Configure passwordless sudo on the remote server, "
                    "or use a root user."
        )

    # /etc/environment: read-modify-write
    existing = _read_file(executor, ENV_FILE)
    new_env = _update_env_lines(existing, proxy_config)
    ok1 = _write_file(executor, ENV_FILE, new_env, sudo=True)

    # /etc/profile.d/proxy-switch.sh: dedicated file, safe to overwrite
    profile_content = _make_profile_content(proxy_config)
    ok2 = _write_file(executor, PROFILE_FILE, profile_content, sudo=True)

    if not ok1 or not ok2:
        errs = []
        if not ok1:
            errs.append(f"Failed to write {ENV_FILE}")
        if not ok2:
            errs.append(f"Failed to write {PROFILE_FILE}")
        return Result(success=False, message="; ".join(errs))
    return Result(success=True, message="System proxy configured")


def disable(executor=None) -> Result:
    """Remove proxy lines from /etc/environment and delete profile script."""
    # Early sudo check
    if executor and not executor.has_sudo():
        return Result(
            success=False,
            message="Sudo access required. Configure passwordless sudo on the remote server, "
                    "or use a root user."
        )

    # Remove proxy lines from /etc/environment
    existing = _read_file(executor, ENV_FILE)
    new_env = _update_env_lines(existing, {})  # empty config → no proxy lines
    _write_file(executor, ENV_FILE, new_env, sudo=True)

    # Delete the profile script
    if executor:
        executor.run(f"sudo rm -f {PROFILE_FILE}")
    else:
        import subprocess
        subprocess.run(["sudo", "rm", "-f", PROFILE_FILE])

    return Result(success=True, message="System proxy disabled")


def status(executor=None) -> StatusInfo:
    """Check proxy state from /etc/environment."""
    content = _read_file(executor, ENV_FILE)

    enabled = "http_proxy=" in content or "HTTP_PROXY=" in content
    proxy = None
    for line in content.split("\n"):
        line = line.strip()
        if line.startswith("http_proxy=") or line.startswith("HTTP_PROXY="):
            proxy = line.split("=", 1)[1].strip('"').strip("'")
            break

    return StatusInfo(
        enabled=enabled,
        proxy=proxy,
        config_file=f"{ENV_FILE}, {PROFILE_FILE}",
        notes="Run 'source /etc/profile.d/proxy-switch.sh' or re-login to apply",
    )


def validate(executor=None) -> List[str]:
    """Check that system proxy is applicable."""
    issues = []
    if not detect(executor):
        issues.append("System proxy is not available on this system")
    return issues

File: src-tauri/src/proxy/remote/system_proxy.rs
use crate::models::{ComponentId, ManualStep, OpResult, ProxyConfig, ProxyStatus};
use crate::proxy::ProxyModule;
use crate::ssh::connection::SshSession;

const ENV_FILE: &str = "/etc/environment";
const PROFILE_FILE: &str = "/etc/profile.d/proxy-switch.sh";

const PROXY_KEYS_LOWER: &[&str] = &["http_proxy", "https_proxy", "ftp_proxy", "no_proxy"];
const PROXY_KEYS_UPPER: &[&str] = &["HTTP_PROXY", "HTTPS_PROXY", "FTP_PROXY", "NO_PROXY"];

pub struct SystemProxyModule;

impl ProxyModule for SystemProxyModule {
    fn config_files(&self) -> Vec<String> {
        vec![ENV_FILE.into(), PROFILE_FILE.into()]
    }
    fn manual_steps(&self) -> Vec<(String, Vec<String>)> {
        vec![(
            "Edit system proxy config".into(),
            vec![
                format!("# Edit {}", ENV_FILE),
                format!("sudo nano {}", ENV_FILE),
                String::new(),
                "# Add these lines:".to_string(),
                "http_proxy=\"http://your-proxy:port\"".into(),
                "https_proxy=\"http://your-proxy:port\"".into(),
                "HTTP_PROXY=\"http://your-proxy:port\"".into(),
                "HTTPS_PROXY=\"http://your-proxy:port\"".into(),
                "no_proxy=\"localhost,127.0.0.1,::1\"".into(),
                "NO_PROXY=\"localhost,127.0.0.1,::1\"".into(),
                String::new(),
                "# Then apply:".to_string(),
                format!("source {}", PROFILE_FILE),
                "# or re-login".to_string(),
            ],
        )]
    }
    fn detect(&self, _session: &SshSession) -> bool {
        true
    }
    fn status(&self, session: &SshSession) -> ProxyStatus {
        let env = session.read_file(ENV_FILE);
        let profile = session.read_file(PROFILE_FILE);
        let http = env_value(&env, "http_proxy")
            .or_else(|| env_value(&env, "HTTP_PROXY"))
            .or_else(|| env_value(&profile, "http_proxy"));
        let https = env_value(&env, "https_proxy")
            .or_else(|| env_value(&env, "HTTPS_PROXY"))
            .or_else(|| env_value(&profile, "https_proxy"));
        let no_proxy = env_value(&env, "no_proxy")
            .or_else(|| env_value(&env, "NO_PROXY"))
            .or_else(|| env_value(&profile, "no_proxy"));
        ProxyStatus {
            component: ComponentId::SystemProxy,
            installed: true,
            enabled: http.is_some() || https.is_some(),
            current_http_proxy: http,
            current_https_proxy: https,
            current_no_proxy: no_proxy,
            current_mirror: None,
            config_files: self.config_files(),
            manual_setup_steps: self
                .manual_steps()
                .into_iter()
                .map(|(t, c)| ManualStep {
                    title: t,
                    commands: c,
                })
                .collect(),
        }
    }
    fn enable(&self, session: &SshSession, config: &ProxyConfig) -> OpResult {
        if !session.has_sudo() {
            return OpResult::err("Sudo access required. Configure passwordless sudo on the remote server.");
        }
        let env_content = update_env_lines(&session.read_file(ENV_FILE), config);
        if let Err(e) = session.write_file(ENV_FILE, &env_content, true) {
            return OpResult::err(format!("Failed to write {}: {}", ENV_FILE, e));
        }
        let profile_content = make_profile_content(config);
        if let Err(e) = session.write_file(PROFILE_FILE, &profile_content, true) {
            return OpResult::err(format!("Failed to write {}: {}", PROFILE_FILE, e));
        }
        OpResult::ok("System proxy configured")
    }
    fn disable(&self, session: &SshSession) -> OpResult {
        if !session.has_sudo() {
            return OpResult::err("Sudo access required.");
        }
        let env_content = update_env_lines(&session.read_file(ENV_FILE), &ProxyConfig::default());
        session.write_file(ENV_FILE, &env_content, true).ok();
        session.run(&format!("sudo rm -f {}", PROFILE_FILE), 10);
        OpResult::ok("System proxy disabled")
    }
}

fn update_env_lines(existing: &str, config: &ProxyConfig) -> String {
    let proxy_prefixes: Vec<&str> = PROXY_KEYS_LOWER
        .iter()
        .map(|k| *k)
        .chain(PROXY_KEYS_UPPER.iter().map(|k| *k))
        .collect();
    let keep: Vec<&str> = existing
        .lines()
        .filter(|line| {
            let trimmed = line.trim();
            if trimmed.starts_with("# Managed by proxy-switch") {
                return false;
            }
            for prefix in &proxy_prefixes {
                if trimmed.starts_with(&format!("{}=", prefix)) {
                    return false;
                }
            }
            true
        })
        .collect();
    let mut new_lines: Vec<String> = keep.iter().map(|s| s.to_string()).collect();
    while new_lines.last().map_or(false, |l| l.is_empty()) {
        new_lines.pop();
    }
    if !new_lines.is_empty() && !new_lines.last().unwrap().is_empty() {
        new_lines.push(String::new());
    }
    new_lines.push("# Managed by proxy-switch".into());
    for key in PROXY_KEYS_LOWER {
        let val = match *key {
            "http_proxy" => &config.http_proxy,
            "https_proxy" => &config.https_proxy,
            "ftp_proxy" => "",
            "no_proxy" => &config.no_proxy,
            _ => "",
        };
        if !val.is_empty() {
            new_lines.push(format!("{}=\"{}\"", key, val));
        }
    }
    for key in PROXY_KEYS_UPPER {
        let val = match *key {
            "HTTP_PROXY" => &config.http_proxy,
            "HTTPS_PROXY" => &config.https_proxy,
            "FTP_PROXY" => "",
            "NO_PROXY" => &config.no_proxy,
            _ => "",
        };
        if !val.is_empty() {
            new_lines.push(format!("{}=\"{}\"", key, val));
        }
    }
    new_lines.push(String::new());
    new_lines.join("\n")
}

fn env_value(content: &str, key: &str) -> Option<String> {
    let prefix = format!("{key}=");
    let export_prefix = format!("export {key}=");
    for line in content.lines() {
        let trimmed = line.trim();
        let rest = if let Some(r) = trimmed.strip_prefix(&export_prefix) {
            Some(r)
        } else if let Some(r) = trimmed.strip_prefix(&prefix) {
            Some(r)
        } else {
            None
        };
        if let Some(rest) = rest {
            let val = rest.trim().trim_matches('"').trim_matches('\'').to_string();
            if !val.is_empty() {
                return Some(val);
            }
        }
    }
    None
}

fn make_profile_content(config: &ProxyConfig) -> String {
    let mut lines = vec!["# Managed by proxy-switch".to_string(), String::new()];
    let keys = [
        ("http_proxy", &config.http_proxy),
        ("https_proxy", &config.https_proxy),
        ("no_proxy", &config.no_proxy),
    ];
    for (key, val) in &keys {
        if !val.is_empty() {
            lines.push(format!("export {}=\"{}\"", key, val));
        }
    }
    lines.push(String::new());
    lines.join("\n")
}
